cmd/relay/events/consumer.go

```diff
@@ -0,0 +1,372 @@
+	// Wait on all limiters before calling the next handler
+	for _, lim := range rsc.limiters {
+		if err := waitForLimiter(ctx, lim); err != nil {
+			return err
+		}
+	}
+	return rsc.Next(ctx, xev)
+}
+
+type instrumentedReader struct {
+	r            io.Reader
+	addr         string
+	bytesCounter prometheus.Counter
+}
+
+func (sr *instrumentedReader) Read(p []byte) (int, error) {
+	n, err := sr.r.Read(p)
+	sr.bytesCounter.Add(float64(n))
+	return n, err
+}
+
+// HandleRepoStream
+// con is source of events
+// sched gets AddWork for each event
+// log may be nil for default logger
+func HandleRepoStream(ctx context.Context, con *websocket.Conn, sched Scheduler, log *slog.Logger) error {
+	if log == nil {
+		log = slog.Default().With("system", "events")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	defer sched.Shutdown()
+
+	remoteAddr := con.RemoteAddr().String()
+
+	go func() {
+		t := time.NewTicker(time.Second * 30)
+		defer t.Stop()
+		failcount := 0
+
+		for {
+
+			select {
+			case <-t.C:
+				if err := con.WriteControl(websocket.PingMessage, []byte{}, time.Now().Add(time.Second*10)); err != nil {
+					log.Warn("failed to ping", "err", err)
+					failcount++
+					if failcount >= 4 {
+						log.Error("too many ping fails", "count", failcount)
+						con.Close()
+						return
+					}
+				} else {
+					failcount = 0 // ok ping
+				}
+			case <-ctx.Done():
+				con.Close()
+				return
+			}
+		}
+	}()
+
+	con.SetPingHandler(func(message string) error {
+		err := con.WriteControl(websocket.PongMessage, []byte(message), time.Now().Add(time.Second*60))
+		if err == websocket.ErrCloseSent {
+			return nil
+		} else if e, ok := err.(net.Error); ok && e.Temporary() {
+			return nil
+		}
+		return err
+	})
+
+	con.SetPongHandler(func(_ string) error {
+		if err := con.SetReadDeadline(time.Now().Add(time.Minute)); err != nil {
+			log.Error("failed to set read deadline", "err", err)
+		}
+
+		return nil
+	})
+
+	lastSeq := int64(-1)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		mt, rawReader, err := con.NextReader()
+		if err != nil {
+			return fmt.Errorf("con err at read: %w", err)
+		}
+
+		switch mt {
+		default:
+			return fmt.Errorf("expected binary message from subscription endpoint")
+		case websocket.BinaryMessage:
+			// ok
+		}
+
+		r := &instrumentedReader{
+			r:            rawReader,
+			addr:         remoteAddr,
+			bytesCounter: bytesFromStreamCounter.WithLabelValues(remoteAddr),
+		}
+
+		var header EventHeader
+		if err := header.UnmarshalCBOR(r); err != nil {
+			return fmt.Errorf("reading header: %w", err)
+		}
+
+		eventsFromStreamCounter.WithLabelValues(remoteAddr).Inc()
+
+		switch header.Op {
+		case EvtKindMessage:
+			switch header.MsgType {
+			case "#commit":
+				var evt comatproto.SyncSubscribeRepos_Commit
+				if err := evt.UnmarshalCBOR(r); err != nil {
+					return fmt.Errorf("reading repoCommit event: %w", err)
+				}
+
+				if evt.Seq < lastSeq {
```

Comment:
I think it should be  `evt.Seq <= lastSeq`, and if that happens the message should just be dropped, right? we don't want to get in a loop re-processing stale events if the upstream accidentally re-emits them?